contrib/python/src/python/pants/contrib/python/checks/tasks/checkstyle/checkstyle.py

```diff
@@ -155,30 +156,29 @@ def checkstyle(self, interpreter, sources):
         failure_count = checker.run(args=args,
                                     stdout=workunit.output('stdout'),
                                     stderr=workunit.output('stderr'))
-        if failure_count > 0 and self.get_options().fail:
-          raise TaskError('{} Python Style issues found. You may try `./pants fmt <targets>`'
-                          .format(failure_count))
         return failure_count
 
   def execute(self):
     """"Run Checkstyle on all found non-synthetic source files."""
-
-    # If we are linting for Python 3, skip lint altogether.
-    # Long-term Python 3 linting solution tracked by:
-    # https://github.com/pantsbuild/pants/issues/5764
-    interpreter = self.context.products.get_data(PythonInterpreter)
-    # Check interpreter is not 'None' for test cases that do not
-    # run the python interpreter selection task.
-    if interpreter and interpreter.version >= (3, 0, 0):
-      self.context.log.info('Linting is currently disabled for Python 3 targets.\n '
-                            'See https://github.com/pantsbuild/pants/issues/5764 for '
-                            'long-term solution tracking.')
-      return
+    python_tgts = self.context.targets(
+      lambda tgt: isinstance(tgt, (PythonTarget))
+    )
+    interpreter_cache = PythonInterpreterCache(PythonSetup.global_instance(),
+                                               PythonRepos.global_instance(),
+                                               logger=self.context.log.debug)
+    tgts_by_compatibility, _ = interpreter_cache.partition_targets_by_compatibility(python_tgts)
 
     with self.invalidated(self.get_targets(self._is_checked)) as invalidation_check:
-      sources = self.calculate_sources([vt.target for vt in invalidation_check.invalid_vts])
-      if sources:
-        return self.checkstyle(interpreter, sources)
+      failure_count = 0
+      tgts_by_compatibility, _ = interpreter_cache.partition_targets_by_compatibility([vt.target for vt in invalidation_check.invalid_vts])
+      for _, targets in tgts_by_compatibility.items():
+        sources = self.calculate_sources(targets)
+        if sources:
+          interpreter = interpreter_cache.select_interpreter_for_targets(targets)
```

Comment:
The key for the `tgts_by_compatibility` dict should already be enough to select an interpreter, I think? This line will re-compute the compatibility of the targets. Not a big deal, but.